simpl/middleware/errors.py

```diff
@@ -0,0 +1,192 @@
+# Copyright (c) 2011-2015 Rackspace US, Inc.
+# All Rights Reserved.
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""Middleware for errors and exceptions.
+
+Example:
+
+    # Allow this middleware to catch, handle and format all errors.
+
+    import bottle
+    from simpl.middleware import errors
+
+    app = bottle.default_app()
+    app.catchall = False
+    # To catch errors that occur in other middleware...
+    app = FormatExceptionMiddleware(app)
+    # Tell bottle to use rest.format_error_response when
+    # handling its own bottle.HTTPErrors
+    app.default_error_handler = rest.format_error_response
+    bottle.run(app=app)
+
+
+    Example responses (the following use catchall=False)
+    -----------------------------------------------------
+
+    An unexpected error if the server is in debug mode:
+
+        # raise TypeError('Poor type choice.')
+        {
+            "code": 500,
+            "exception": "TypeError('Poor type choice.',)",
+            "message": "We're sorry, something went wrong.",
+            "reason": "Internal Server Error",
+            "traceback": "Traceback (most recent call last): ... ",
+        }
+
+    An "expected" error if the server is in debug mode:
+
+        # raise SimplHTTPError(body='Hey!', status=405)
+        {
+            "code": 405,
+            "exception": "SimplHTTPError('Hey!',)",
+            "message": "Hey!",
+            "reason": "Method Not Allowed",
+            "traceback": "Traceback (most recent call last): ..."
+        }
+
+    Unexpected error without debug mode:
+
+        # raise TypeError('Poor type choice.')
+        {
+            "code": 500,
+            "message": "We're sorry, something went wrong.",
+            "reason": "Internal Server Error"
+        }
+
+    An "expected" error without debug mode:
+
+        # raise SimplHTTPError(body='Hey!', status=405)
+        {
+            "code": 405,
+            "message": "Hey!",
+            "reason": "Method Not Allowed"
+        }
+"""
+from __future__ import print_function
+
+import logging
+import sys
+import traceback as tb_mod
+
+import bottle
+
+from simpl import exceptions as simpl_exc
+from simpl import rest
+
+LOG = logging.getLogger(__name__)
+
+
+def _catchall_enabled(app):
+    """Check the bottle app for catchall."""
+    while hasattr(app, 'app'):
```

Comment:
How about checking bottle.default_app() as a failsafe if following th chain doesn't get you to a bottle instance?